ext/random/random.c

```diff
@@ -0,0 +1,1488 @@
+	php_random_numbergenerator_state_combinedlcg *s = (php_random_numbergenerator_state_combinedlcg *)state;
+	zval tmp;
+	int i;
+
+	for (i = 0; i < 2; i++) {
+		ZVAL_LONG(&tmp, s->s[i]);
+		zend_hash_next_index_insert(data, &tmp);
+	}
+	ZVAL_LONG(&tmp, s->seeded);
+	zend_hash_next_index_insert(data, &tmp);
+
+	return SUCCESS;
+}
+
+static int combinedlcg_unserialize(void *state, HashTable *data) {
+	php_random_numbergenerator_state_combinedlcg *s = (php_random_numbergenerator_state_combinedlcg *)state;
+	zval *tmp;
+	int i;
+
+	for (i = 0; i < 2; i++) {
+		tmp = zend_hash_index_find(data, i);
+		if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+			return FAILURE;
+		}
+		s->s[i] = Z_LVAL_P(tmp);
+	}
+	tmp = zend_hash_index_find(data, 2);
+	if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+		return FAILURE;
+	}
+	s->seeded = Z_LVAL_P(tmp);
+
+	return SUCCESS;
+}
+
+static zend_object *php_random_numbergenerator_combinedlcg_new(zend_class_entry *ce) {
+	return php_random_numbergenerator_common_init(ce, &php_random_numbergenerator_algo_combinedlcg, &random_numbergenerator_combinedlcg_object_handlers);
+}
+
+/* CombinedLCG end */
+
+/* Secure begin */
+
+static uint64_t secure_generate(void *state) {
+	uint64_t ret;
+
+	php_random_bytes_silent(&ret, sizeof(uint64_t));
+
+	return ret;
+}
+
+static zend_object *php_random_numbergenerator_secure_new(zend_class_entry *ce) {
+	return php_random_numbergenerator_common_init(ce, &php_random_numbergenerator_algo_secure, &random_numbergenerator_secure_object_handlers);
+}
+
+/* Secure end */
+
+/* User begin */
+
+static uint64_t user_generate(void *state) {
+	php_random_numbergenerator_state_user *s = (php_random_numbergenerator_state_user *)state;
+	zval retval;
+
+	zend_call_known_instance_method_with_0_params(s->method, s->object, &retval);
+
+	return (uint64_t) Z_LVAL(retval);
+}
+
+/* User end */
+
+const php_random_numbergenerator_algo php_random_numbergenerator_algo_xorshift128plus = {
+	sizeof(uint64_t),
+	sizeof(php_random_numbergenerator_state_xorshift128plus),
+	xorshift128plus_generate,
+	xorshift128plus_seed,
+	xorshift128plus_serialize,
+	xorshift128plus_unserialize
+};
+
+const php_random_numbergenerator_algo php_random_numbergenerator_algo_mersennetwister = {
+	sizeof(uint32_t),
+	sizeof(php_random_numbergenerator_state_mersennetwister),
+	mersennetwister_generate,
+	mersennetwister_seed,
+	mersennetwister_serialize,
+	mersennetwister_unserialize
+};
+
+const php_random_numbergenerator_algo php_random_numbergenerator_algo_combinedlcg = {
+	sizeof(int32_t),
+	sizeof(php_random_numbergenerator_state_combinedlcg),
+	combinedlcg_generate,
+	combinedlcg_seed,
+	combinedlcg_serialize,
+	combinedlcg_unserialize
+};
+
+const php_random_numbergenerator_algo php_random_numbergenerator_algo_secure = {
+	sizeof(uint64_t),
+	0,
+	secure_generate,
+	NULL,
+	NULL,
+	NULL
+};
+
+const php_random_numbergenerator_algo php_random_numbergenerator_algo_user = {
+	sizeof(zend_long),
```

Comment:
If I am reading this correctly, it is not possible to simple decorate a mersenne twister number (or lcg) generator in userland and forward the generated values 1:1 in the `generate()` function, due to the different sizes?
I think that information should be exposed on the `NumberGenerator` interface, to avoid this issue.

Reply:
Or, as an alternative, the `generate()` function could return byte strings, which then can be trivially used thanks to the natively encoded length into a string.

Reply:
Yes. It is currently impossible.
We have already considered them (32-bit RNGs) to be unnecessary and have not provided the means to do so, but if you think they are necessary, we would be happy to hear your opinion in the In ML.

The idea of using byte strings as seed/generated values has been mentioned in ML, but I'm thinking it's too subtle for both performance and practicality.

If this problem is to be addressed, it would probably be better to add method to get the generate size.

Reply:
I was noting that here because the RFC did not attempt to describe that fact and I only discovered this in code.

I'm not sure what the best way is, but at least the RFC should note and discuss this, so that on ML there can be further discussion around that.

Reply:
Thank you. I will revise the RFC.

Reply:
RFC updated
https://wiki.php.net/rfc/rng_extension

Reply:
This problem has been resolved by changes in the implementation.
https://github.com/colopl/php-src/blob/upstream_rfc/scoped_rng_for_pr/ext/random/tests/engine/user_compatibility.phpt

Reply:
thanks :-)